embeddings/hybrid-search.ipynb

```diff
@@ -0,0 +1,1124 @@
+    "\n",
+    "In JSON, the data file looks like this:\n",
+    "\n",
+    "```\n",
+    "{\"id\": \"3\", \"sparse_embedding\": {\"values\": [0.1, 0.2], \"dimensions\": [1, 4]}}\n",
+    "{\"id\": \"4\", \"sparse_embedding\": {\"values\": [-0.4, 0.2, -1.3], \"dimensions\": [10, 20, 20]}}\n",
+    "```\n",
+    "\n",
+    "Each item should have a `sparse_embedding` property that has `values` and `dimensions` properties. Sparse embeddings have thousands of dimensions with a few non-zero values. This data format works efficiently because it contains the non-zero values only with their positions in the space.\n",
+    "\n",
+    "As a sample dataset, we'll use the [Google Merch Shop](https://shop.merch.google/) dataset, which has about 200 rows of Google-branded goods."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "KAa1Ff5zaNKf",
+    "outputId": "6e17f43a-d11f-462c-bf27-1402a7abf035"
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "0                          Google Sticker\n",
+       "1                    Google Cloud Sticker\n",
+       "2                       Android Black Pen\n",
+       "3                   Google Ombre Lime Pen\n",
+       "4                    For Everyone Eco Pen\n",
+       "                      ...                \n",
+       "197        Google Recycled Black Backpack\n",
+       "198    Google Cascades Unisex Zip Sweater\n",
+       "199    Google Cascades Womens Zip Sweater\n",
+       "200         Google Cloud Skyline Backpack\n",
+       "201       Google City Black Tote Backpack\n",
+       "Name: title, Length: 202, dtype: object"
+      ]
+     },
+     "execution_count": 4,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import pandas as pd\n",
+    "\n",
+    "CSV_URL = \"https://raw.githubusercontent.com/GoogleCloudPlatform/generative-ai/main/gemini/sample-apps/photo-discovery/ag-web/google_merch_shop_items.csv\"\n",
+    "\n",
+    "# Load the CSV file into a DataFrame\n",
+    "df = pd.read_csv(CSV_URL)\n",
+    "df[\"title\"]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "AkcwG37lAfoS"
+   },
+   "source": [
+    "## How to use Token-based Search\n",
+    "\n",
+    "With the dataset, we will create sparse embeddings for implementing a token-based search with Vector Search."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "UBzaDGpYYodr"
+   },
+   "source": [
+    "### Prepare a TF-IDF vectorizer\n",
+    "With this dataset, we'll train a vectorizer, a model that generates sparse embeddings from a text. This example uses [TfidfVectorizer in scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html), which is a basic vectorizer that uses the [TF-IDF algorithm](https://en.wikipedia.org/wiki/Tf%E2%80%93idf)."
```

Comment:
I believe common libraries like `pandas` and `sklearn` are already included in the Colab environment, but not 100% sure on the others such as Colab Enterprise. Be sure to test and add these libraries to the top `pip install` cell if needed.

Reply:
The notebook is tested with Colab Ent, and also Workbench also includes pandas and sklearn.